content/en/blog/_posts/2021-06-08-Kubeadm-Avoid-Default-Address.md

```diff
@@ -0,0 +1,285 @@
+  Node's status.
+
+This post describes work done with version 1.21.1 of kubeadm and
+kubernetes.
+
+You may also like to read a related blog post, [Kubernetes Setup Using
+Ansible and
+Vagrant](https://kubernetes.io/blog/2019/03/15/kubernetes-setup-using-ansible-and-vagrant/).
+In contrast, the post at hand is about comprehensively overriding the
+default address and not being specific to one form of automation.
+
+As an added bonus, I will also mention how to stop DHCP from assigning
+the same address to every node for networks where that is not desired.
+This comes first because everything else depends on it; you can skip
+that section if you are not having that problem.
+
+## Getting DHCP to assign unique addresses
+
+This is a problem that I had when using a common Ubuntu 20.04 snapshot
+to base my various VMs on.  This snapshot includes a file named
+`/etc/machine-id`, and so all my VMs get the same content in that file
+(unless I take extra steps).  This is relevant because in Ubuntu
+20.04, the default DHCP client is `systemd.networkd`, and its default
+configuration causes it to send an explicit "DHCP client identifier"
+(in addition to the client MAC address) in each request and that
+identifier is derived from `/etc/machine-id` and _not_ the relevant
+network interface's MAC address.  When an explicit client ID is
+supplied, the DHCP server uses only that --- not also the client MAC
+address.  Howevever, rather than having to tweak each derived VM, I
+tweaked the common snapshot to make the DHCP client eschew that
+explicit client ID --- leaving the DHCP server to look only at the
+client's MAC address.  The relevant bit of `systemd.networkd`
+configuration is normally driven by a higher level tool called
+`netplan`.  I tweaked its configuration by creating a drop-in file
+named `/etc/netplan/10-enp0s8.yaml` (because `enp0s8` is the name of
+the network interface that I want to get a distinct IPv4 address on
+each node) with the following contents.
+
+```yaml
+network:
+  ethernets:
+    enp0s8:
+      dhcp4: true
+      dhcp-identifier: mac
+  version: 2
+```
+
+## The kube-apiserver's address in its Endpoints object
+
+The default for this can be overridden by passing
+`--advertise-address=${desired_address}` on the kube-apiservers's
+command line.  You can get `kubeadm init` to do that either with a
+command line flag or a bit of configuration file content.  The command
+line flag would be `--apiserver-advertise-address=192.168.56.101`,
+assuming the desired address for the control node is 192.168.56.101.
+I will use that example address throughout this post.
+
+I used a configuration file, named `ka-init.conf`.  The following
+content will get this job done.
+
+```yaml
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+apiServer:
+  extraArgs:
+    advertise-address: 192.168.56.101
+```
+
+## The Subject Alternative Name list in the kube-apiserver's certificate
```

Comment:
I'd actually write:
```suggestion
## The subjectAltName list in the kube-apiserver's certificate
```

Reply:
Why?  I think "Subject Alternative Name" is the canonical jargon.  That is what `openssl x509 -text` prints, and I find it in many articles about these certificates.